trl/trainer/grpo_trainer.py

```diff
@@ -316,9 +324,43 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
             prompt_inputs["input_ids"] = prompt_inputs["input_ids"][:, -self.max_prompt_length :]
             prompt_inputs["attention_mask"] = prompt_inputs["attention_mask"][:, -self.max_prompt_length :]
 
-        # Generate completions
-        with unwrap_model_for_generation(model, self.accelerator) as unwrapped_model:
-            prompt_completion_ids = unwrapped_model.generate(**prompt_inputs, generation_config=self.generation_config)
+        # Generate completions using either vLLM or regular generation
+        if self.args.use_vllm:
+            # Gather all prompts from all processes to the main process
+            all_prompts = self.accelerator.gather(prompts)
+            all_prompt_inputs = self.accelerator.gather(prompt_inputs["input_ids"])
+
+            if self.accelerator.is_main_process:
+                if self.state.global_step != self._last_loaded_step:
+                    self.vllm_client.load_weights(model.state_dict())
+                    self._last_loaded_step = self.state.global_step
+
+                # Get completions from vLLM for all prompts
+                completions_data = self.vllm_client.chat(all_prompts)
+                completions = completions_data["completions"]
```

Comment:
@kashif here we need an if/else, depending on whether the data is conversational. But I'm wondering if we can have this if/else directly in the client side, and use a unique `generation`. In other words:

what we should do here:

```python
if is_conversational:
    completions_data = self.vllm_client.chat(all_prompts)
    ...
else:
   completions_data = self.vllm_client.generate(all_prompts)
   ...
```

OR, modify the client, so that here it's handling both cases smoothly:

```python
completions_data = self.vllm_client.generate(all_prompts)  # would work for both cons and not conv
```

wdyt? IMO the later is the best.

Reply:
An alternative would be to use `apply_chat_template` and pass that directly to `generate()`. It's worth pointing out there's some bugs with how BOS tokens are handled on these endpoints: https://github.com/vllm-project/vllm/issues/9519

Reply:
That's a good point! Let's try not to rely on vllm to apply chat template then